man/ibparsefabricconf.8.in
.TH IBPARSEFABRICCONF 8 "Sept 29, 2010" "Pragmatic Infiniband Utilities" "Infiniband Diagnostics"

.SH NAME
ibparsefabricconf \- parse fabricconf file

.SH SYNOPSIS
.B ibparsefabricconf [options] [node] [port]

.SH DESCRIPTION
.PP
ibparsefabricconf reads the ib fabric config file and reports the links configured
there.  If node is specified print the config for that node.
If port is specified further limit the output to that single node/port.

.SH OPTIONS

.PP
.TP
\fB\-config, \-c <ibfabricconf>\fR
Specify an alternate config file (default: @PIU_CONFIG_PATH@/pragmaticIB/ibfabricconf.xml)

.TP
\fB\-\-warn_dup, \-w\fR
print warning about duplicate links found in fabric config

.TP
\fB\-\-check_dup\fR
only print duplicates links found in fabric config.  This is useful for
verifing your config file.

.TP
\fB\-\-delim_out, \-d <deliminator>\fR
output colums deliminated by <deliminator>

.SH AUTHOR
.TP
Ira Weiny
.RI < weiny2@llnl.gov >


ChangeLog
- Thu Oct 28 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.1.4-2
	* : Add sierra-force-QDR.sh
	* : Add ibmad_renice for sierra

- Thu Oct 18 2010 Ira Weiny <weiny2@llnl.gov>

	* : ibtraceroute: update help message

- Thu Oct 07 2010 Ira Weiny <weiny2@llnl.gov>

	* : ibtraceroute: 
		  update ibtraceroute with multicast functionality 
		  also default the "from" node to the current running node

- Thu Oct 07 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.1.4-1
	* : ibgraphfabric.c: fix partial scan of fabric

- Thu Oct 07 2010 Albert Chu <chu11@llnl.gov>

	* : support comma separated output in lfts_sort.pl
	* : support 36 or 24 port switches in lfts_sort.pl

- Fri Sep 10 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.1.3-1
	* : ibhcacounters.8.in: Add man page
	* : ibhcacounters: Add card name to output

- Thu Sep 09 2010 Ira Weiny <weiny2@llnl.gov>

	* : ibsrp.8.in: fixed config file location
	* : ibtrackerrors.8.in: Add ibtrackerrors man page
	* : ibtrackerrors.in: Fix help message
	* : ibtrackerrors.in: Add config file support
	* : pragmatic-IB.conf: Add config file

- Wed Sep 08 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.1.2-2
	* : ibcheckfabric.in: fix incorrect reporting
	* : ibtrackerrors.in: Allow for correct error names from spec

- Mon Aug 23 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.1.2-1
	* : ibtrackerrors.in: add stderr output to log file.

- Thu Aug 19 2010 Ira Weiny <weiny2@llnl.gov>

	* : ibtrackerrors.in: allow additional erros to be suppressed
	* : ibhcacounters: remove message which says this only works on ConnectX

- Fri Aug 06 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.1.1-1
	* : ibtrackerrors: Add support to limit the scan to switches only

- Mon Aug 02 2010 Albert Chu <chu11@llnl.gov>

	* : add check_wiring_sierra.pl

- Tue Jul 29 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.1.0-2
	* : ibnodesinmcast: Allow for consolidated MLID's without consolidated MGID's
	* : ibtraceroute: use absolute path for rdma_cm_query

- Tue Jul 20 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.1.0-1
	* : ibgraphfabric: Convert to libibnetdisc ABI version 4
	* : slurm_topology: Convert to libibnetdisc ABI version 4

- Tue Jul 20 2010 Al Chu <chu11@llnl.gov>

	* : fix typos

- Wed Jul 14 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.0.6-1

- Fri Jul 09 2010 Ira Weiny <weiny2@llnl.gov>

	* : ibhcacounters: Add support for QIB cards

- Fri Jul 09 2010 Ira Weiny <weiny2@llnl.gov>

	* : ibcheckverbs: explicitly call ibv_rc_pingpong
	* : ibcheckverbs: explicitly call pdsh
	* : ibcheckverbs: add support for renamed ib_rdma_bw tool

- Fri Jun 25 2010 Ira Weiny <weiny2@llnl.gov>

	* : ibsrp.in: fix issue of script killing opensm process which were not
	    owned by the script.  Also include stdout/stderr in log file.

- Mon May 17 2010 Ira Weiny <weiny2@llnl.gov>

	* : ibcheckverbs: clean up output

- Tue May 04 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.0.5-1
	* : Integrate Hostlist.pm as PIUHostlist.pm

- Mon May 03 2010 Ira Weiny <weiny2@llnl.gov>

	* : Add whatsup check to ibnodesinmcast
	* : Add pdsh check to ibcheckverbs
	* : Update and add man pages (ibcheckverbs, slurm_topology)
	* : ibgraphfabric.pl: Remove in favor of ibgraphfabric
	* : ibbouncelinks.pl: Remove obsolete "-R" option
			Update docs
			increase timeout for QDR links
	* : ibcheckfabric: Fix issue with new iblinkinfo output

- Mon Apr 26 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.0.4-4
	* : Update all the autoconf stuff to build without autoconf
	* : Check in "configure" support

- Fri Apr 23 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.0.4-3

- Fri Apr 23 2010 Mark Grondona <grondona1@llnl.gov>

	* : Use genders_query correctly.

- Thu Apr 22 2010 Ira Weiny <weiny2@llnl.gov>

	* : Update hostlist code to latest version

- Wed Apr 07 2010 Ira Weiny <weiny2@llnl.gov>

	* : ibtraceroute: clean up error when no nodes are specified
	* : ibtraceroute: add "-m" option
	* : ibtraceroute: Default one node to "local" node
	* : ibtraceroute: fixed exit codes on lookup failure.

- Wed Mar 31 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.0.4-2

- Tue Mar 30 2010 Ira Weiny <weiny2@llnl.gov>

	* : Added --hosts option to slurm_topology
	* : cleaned up output of slurm_topology
		 - ordered switches
		 - grouped levels
		 - printed missing and looked for "compute" nodes

- Fri Feb 19 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.0.4
	* : Added ibhcacounters

- Thu Feb 18 2010 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.0.3-2

- Tue Feb 09 2010 Ira Weiny <weiny2@llnl.gov>

	* : Added -g/-q option to slurm_topology

- Thu Feb 04 2010 Ira Weiny <weiny2@llnl.gov>

	* : Added -n option to slurm_topology

- Tue Dec 10 2009 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.0.3
	* : change ibsrp log dir to "/var/log/ibsrp"
	* : Remove modprobing of IB cards from ibsrp init script.  depmod and
	    modprobe should deal with "modprobe ib_srp" correctly.
	* : changed the unloading of ib_srp module and the stopping of the daemons

- Tue Nov 17 2009 Ira Weiny <weiny2@llnl.gov>

	* : tag 1.0.2
	* : Initial checkin of ibgraphfabric C version.
	* : Remove unused assignment in slurm_topology

- Tue Oct 27 2009 Ira Weiny <weiny2@llnl.gov>

	* : bump version to 1.0.1 as slurm_topology now requires libibnetdisc
	* : Add slurm_topology utility

- Fri Oct 16 2009 Ira Weiny <weiny2@llnl.gov>

	* : improve ibbouncelinks.pl
		Add QDR skip option. (including lengthening timeout for QDR ports)
		Update help message
		pretify output with progress "."

- Thu Oct 08 2009 Ira Weiny <weiny2@llnl.gov>

	* : add device and port support to node specification in ibcheckverbs

- Fri Oct 02 2009 Ira Weiny <weiny2@llnl.gov>

	* : add -m option to ibtrackerrors and ibcheckfabric

- Wed Sep 30 2009 Ira Weiny <weiny2@llnl.gov>

	* : add support for --clear-errors in ibqueryerrors to ibtrackerrors 

- Tue Sep 22 2009 Ira Weiny <weiny2@llnl.gov>

	* : Add experimental ibcreateswitchmap.sh
	* : Fix check_wiring_coastal.pl

- Tue Aug 25 2009 Ira Weiny <weiny2@llnl.gov>


src/ibparsefabricconf.c
/*
 * Copyright (C) 2010 Lawrence Livermore National Security
 * Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 * Written by Ira Weiny weiny2@llnl.gov
 * UCRL-CODE-235440
 * 
 * This file is part of pragmatic-infiniband-tools (PIU), useful tools to manage
 * Infiniband Clusters.
 * For details, see http://www.llnl.gov/linux/.
 * 
 * PIU is free software; you can redistribute it
 * and/or modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 * 
 * PIU is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with
 * PIU; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <getopt.h>
#include <string.h>
#include <iba/ib_types.h>
#include <infiniband/ibfabricconf.h>

char *fabricconf_file = NULL;
char *argv0 = NULL;

char *delim_out = NULL;

/** =========================================================================
 * Borrow from ibnetdiscover
 */
char *linkspeed_str(int speed)
{
	switch (speed) {
	case IB_LINK_SPEED_ACTIVE_2_5:
		return ("2.5 Gbps");
		//return ("SDR");
		break;
	case IB_LINK_SPEED_ACTIVE_5:
		return ("5.0 Gbps");
		//return ("DDR");
		break;
	case IB_LINK_SPEED_ACTIVE_10:
		return ("10.0 Gbps");
		//return ("QDR");
		break;
	}
	return ("???");
}

char *linkwidth_str(int width)
{
	switch (width) {
	case IB_LINK_WIDTH_ACTIVE_1X:
		return ("1X");
		break;
	case IB_LINK_WIDTH_ACTIVE_4X:
		return ("4X");
		break;
	case IB_LINK_WIDTH_ACTIVE_8X:
		return ("8X");
		break;
	case IB_LINK_WIDTH_ACTIVE_12X:
		return ("12X");
		break;
	}
	return ("??");
}

void
print_port(ibfc_port_t *port, void *user_data)
{
	char prop[256];
	if (delim_out)
		printf("%s%s%d%s%d%s%s%s%s%s%s\n",
			ibfc_port_get_name(port),
			delim_out,
			ibfc_port_get_port_num(port),
			delim_out,
			ibfc_port_get_port_num(ibfc_port_get_remote(port)),
			delim_out,
			ibfc_port_get_name(ibfc_port_get_remote(port)),
			delim_out,
			linkspeed_str(ibfc_prop_get_speed(ibfc_port_get_prop(port))),
			delim_out,
			linkwidth_str(ibfc_prop_get_width(ibfc_port_get_prop(port))));
	else
		printf ("\"%30s\" %4d  ==(%s)==>  %4d \"%s\"\n",
			ibfc_port_get_name(port),
			ibfc_port_get_port_num(port),
			ibfc_prop_str(port, prop, 256),
			ibfc_port_get_port_num(ibfc_port_get_remote(port)),
			ibfc_port_get_name(ibfc_port_get_remote(port))
			);
}

/** =========================================================================
 */
static int
usage(void)
{
        fprintf(stderr,
"%s [options] [node] [port]\n"
"Usage: parse the fabricconf file\n"
"\n"
"Options:\n"
"  --config, -c <ibfabricconf> Use an alternate fabric config file (default: %s)\n"
"  --warn_dup, -w print warning about duplicate links found in fabric config\n"
"  --check_dup only print duplicates\n"
"  --delim_out, -d <deliminator> output colums deliminated by <deliminator>\n"
"  [node] if node is specified print ports for that node\n"
"  [port] if port is specified print information for just that port (default \"all\")\n"
"         if neither node nor port is specified print entire config file\n"
"\n"
, argv0,
IBFC_DEF_CONFIG
);
        return (0);
}


int
main(int argc, char **argv)
{
	ibfc_conf_t *fabricconf;
	int rc = 0;
	int warn_dup = 0;
	int check_dup = 0;
        char  ch = 0;
        static char const str_opts[] = "hc:wd:";
        static const struct option long_opts [] = {
           {"help", 0, 0, 'h'},
           {"config", 1, 0, 'c'},
           {"warn_dup", 0, 0, 'w'},
           {"check_dup", 0, 0, 1},
           {"delim_out", 1, 0, 'd'},
	   {0, 0, 0, 0}
        };

	argv0 = argv[0];

        while ((ch = getopt_long(argc, argv, str_opts, long_opts, NULL))
                != -1)
        {
                switch (ch)
                {
			case 'c':
				fabricconf_file = strdup(optarg);
				break;
			case 'w':
				warn_dup = 1;
				break;
			case 1:
				check_dup = 1;
				break;
			case 'd':
				delim_out = strdup(optarg);
				break;
                        case 'h':
                        default:
                        	exit(usage());
                }
	}

	argc -= optind;
	argv += optind;

	fabricconf = ibfc_alloc_conf();
	if (!fabricconf) {
		fprintf(stderr, "ERROR: Failed to alloc fabricconf\n");
		exit(1);
	}

	if (check_dup)
		ibfc_set_stderr(fabricconf, stdout);

	ibfc_set_warn_dup(fabricconf, warn_dup | check_dup);
	rc = ibfc_parse_file(fabricconf_file, fabricconf);

	if (rc) {
		fprintf(stderr, "ERROR: failed to parse fabric config "
			"\"%s\":%s\n", fabricconf_file, strerror(rc));
		return (rc);
	}

	if (check_dup)
		goto done;

	if (delim_out) {
		printf("Fabric Name%s%s\n", delim_out, ibfc_conf_get_name(fabricconf));
		printf("Node%sPort%sRem Port%sRem Node%sSpeed%sWidth\n",
			delim_out, delim_out, delim_out, delim_out, delim_out);
	} else
		printf("Fabric Name: %s\n", ibfc_conf_get_name(fabricconf));

	if (argv[0]) {
		char prop[256];
		int p_num = 1;
		if (argv[1]) {
			p_num = strtol(argv[1], NULL, 0);
			ibfc_port_t *port = ibfc_get_port(fabricconf, argv[0], p_num);
			if (port)
				print_port(port, NULL);
			else
				fprintf (stderr, "ERROR: \"%s\":%d port not found\n",
					argv[0], p_num);
		} else {
			ibfc_port_list_t *port_list;
			int rc = ibfc_get_port_list(fabricconf, argv[0],
							&port_list);
			if (rc) {
				fprintf(stderr, "ERROR: Failed to get port "
					"list for \"%s\":%s\n",
					argv[0], strerror(rc));
			} else {
				ibfc_iter_port_list(port_list, print_port,
							NULL);
				ibfc_free_port_list(port_list);
			}
		}
	} else {
		ibfc_iter_ports(fabricconf, print_port, NULL);
	}

done:
	ibfc_free(fabricconf);

	return (rc);
}

